src/burst_buffer/dws.lua

```diff
@@ -0,0 +1,227 @@
+		ret = false,
+		result = "",
+	},
+}
+DWS.__index = DWS
+
+setmetatable(DWS, {
+	__call = function(cls, ...)
+		local self = setmetatable({}, cls)
+		self:new()
+		return self
+	end,
+})
+
+function DWS:new()
+	self.yaml = self:template()
+	return self
+end
+
+-- DWS.template will return a string containing a template for a Workflow
+-- resource with keywords that must be replaced by the caller.
+function DWS:template()
+	return [[
+apiVersion: dws.cray.hpe.com/v1alpha1
+kind: Workflow
+metadata:
+  name: WF_NAME
+spec:
+  desiredState: "proposal"
+  dwDirectives: []
+  wlmID: "WLMID"
+  jobID: JOBID
+  userID: USERID
+  groupID: GROUPID
+]]
+end
+
+-- DWS.initialize will replace keywords in a workflow template to turn
+-- it into a completed workflow resource.
+function DWS:initialize(wf_name, wlmID, jobID, userID, groupID)
+	self.name = wf_name
+	yaml = self.yaml
+	yaml = string.gsub(yaml, "WF_NAME", wf_name)
+	yaml = string.gsub(yaml, "WLMID", wlmID)
+	yaml = string.gsub(yaml, "JOBID", jobID)
+	yaml = string.gsub(yaml, "USERID", userID)
+	yaml = string.gsub(yaml, "GROUPID", groupID)
+	self.yaml = yaml
+end
+
+-- DWS.save will save the text of the Workflow resource into a file
+-- of the given name.
+-- On success this returns true.
+-- On failure this returns false and an optional error message.
+function DWS:save(fname)
+	local f = io.open(fname, "w")
+	if f == nil 
+	then
+		local msg = "unable to open " .. fname
+		return false, msg
+	end
+	f:write(self.yaml)
+	local rc = {f:close()}
+	-- Success or failure is a boolean in rc[1].
+	return rc[1]
+end
+
+-- DWS.apply will apply the Workflow resource via kubectl.
+-- On success this returns true.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:apply(fname)
+	return self:io_popen("kubectl apply -f " .. fname .. " 2>&1")
+end
+
+-- DWS.delete will delete the named Workflow resource via kubectl.
+-- On success this returns true.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:delete()
+	return self:io_popen("kubectl delete workflow " .. self.name .. " 2>&1")
+end
+
+-- DWS.get_jsonpath will get the specified values from the Workflow resource.
+-- On success this returns true and a string containing the result of the query.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:get_jsonpath(path)
+	local cmd = "kubectl get workflow " .. self.name .. " -o jsonpath='" .. path .. "' 2>&1"
+	return self:io_popen(cmd)
+end
+
+-- DWS.get_current_state will get the status of the Workflow resource with
+-- respect to its desired state.
+-- On success this returns true and an array containing the desiredState,
+-- the current state, and the status of the current state.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:get_current_state()
+	local ret, output = self:get_jsonpath([[{.spec.desiredState}{"\n"}{.status.state}{"\n"}{.status.status}{"\n"}]])
```

Comment:
Instead of passing around the state as a string, can we parse it into a table? This would help a lot with readability. The code easier to read if you get the "desiredState" property from a table instead of having to remember that desiredState is in the first position.

Reply:
Yes, I think that would be better.